File: ReSTServices/RestDemo1/app/src/main/java/edu/cs4730/restdemo1/MainActivity.java
package edu.cs4730.restdemo1;

import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import edu.cs4730.restdemo1.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {
    String TAG = "MainActivity";
    ActivityMainBinding binding;
    myAdapter mAdapter;
    JSONArray list = null;
    URI uri;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.toolbar);

        try {
            uri = new URI("https://jsonplaceholder.typicode.com/posts");
        } catch (URISyntaxException e) {
            Log.wtf(TAG, "error with uri");
            e.printStackTrace();
        }

        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show();
            }
        });

        //setup the RecyclerView

        binding.list.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        binding.list.setItemAnimator(new DefaultItemAnimator());
        //setup the adapter, which is myAdapter, see the code.  set it initially to null
        mAdapter = new myAdapter(null, getApplicationContext());
        //add the adapter to the recyclerview
        binding.list.setAdapter(mAdapter);

        //SwipeRefreshlayout setup.
        //setup some colors for the refresh circle.
        binding.activityMainSwipeRefreshLayout.setColorSchemeResources(R.color.orange, R.color.green, R.color.blue);
        //now setup the swiperefrestlayout listener where the main work is done.
        binding.activityMainSwipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                new Thread(new doNetwork(uri)).start();
            }
        });
        binding.activityMainSwipeRefreshLayout.setRefreshing(true);
        //call the refresh code manually here.
        new Thread(new doNetwork(uri)).start();

    }

    /**
     * simple helper method to update the UI.
     */
    public void mkmsg(String str) {
        //so the client uses a handler, while the server code is going to use runUI method.
        //the method name makes no sense here, but keeping the name, so it matches the tpcclient.
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    list = new JSONArray(str);
                    mAdapter.setData(list);
                } catch (JSONException e) {
                    e.printStackTrace();
                    Toast.makeText(getBaseContext(), "Data response is empty", Toast.LENGTH_SHORT).show();
                }
                binding.activityMainSwipeRefreshLayout.setRefreshing(false);  //turn off the refresh.
            }
        });
    }


    //Simple class that takes an InputStream and return the data
    //as a string, with line separators (ie end of line markers)
    private String readStream(InputStream in) {
        BufferedReader reader = null;
        StringBuilder sb = new StringBuilder("");
        String line = "";
        String NL = System.getProperty("line.separator");
        try {
            reader = new BufferedReader(new InputStreamReader(in));
            while ((line = reader.readLine()) != null) {
                sb.append(line).append(NL);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (reader != null) {
                try {
                    reader.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return sb.toString();
    }


    /**
     * actual thread class that calls executeHpptGet() above.
     */
    class doNetwork implements Runnable {
        URI myUri;

        doNetwork(URI uri) {
            myUri = uri;
        }

        public void run() {
            String page = "";
            try {
                URL myUrl = uri.toURL();
                HttpURLConnection con = (HttpURLConnection) myUrl.openConnection();
                page = readStream(con.getInputStream());
                con.disconnect();
            } catch (Exception e) {
                Log.wtf(TAG, e.getMessage());
                return;
            }
            Log.wtf(TAG, "finished, updating result.");
            mkmsg(page);
        }
    }

}

File: tcpserv_kt/app/src/main/java/edu/cs4730/tcpserv_kt/MainActivity.kt
package edu.cs4730.tcpserv_kt

import android.net.wifi.WifiManager
import android.os.Bundle
import android.text.format.Formatter
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import edu.cs4730.tcpserv_kt.databinding.ActivityMainBinding
import java.io.*
import java.net.ServerSocket

/**
 * Note, https://koz.io/android-m-and-the-war-on-cleartext-traffic/
 * In the AndroidManifest.xml there is < application ... android:usesCleartextTraffic="true" ...
 * Actually this app didn't need it to work, but that is likely to change.
 *
 * If you are running this in emulators, make sure to run
 * adb forward tcp:3012 tcp:3012  for the server.
 * Do this on the first emulator, then launch the second one and run the client on the second emulator.
 * This assumes port 3012, so if you changed it, then change adb line as well.
 *
 * Note, that while the IP address of the server is listed, if you are running on emulator use the host
 * address in tcpclient ip 10.0.2.2 (this is the address used by android for the host system.
 *
 * If you want to talk the server from program like telnet, or whatever.  localhost: 3012  is the address
 *
 * This a simple network server code.  The user must push the make connection button to setup the
 * server to accept connections.
 *
 * Note use adb forward tcp:3012 tcp:3012
 * to setup the emulator to receive.
 */
class MainActivity : AppCompatActivity() {
    lateinit var binding: ActivityMainBinding
    var myNet: Thread? = null
    var TAG = "TCPserv"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.logger.append("\n\n")
        binding.makeconn.setOnClickListener {
            //this way creates the thread anonymously.  quick and dirty, but generally a bad idea.
            //Thread(doNetwork()).start()

            //better way is this way, where we have access to the thread variable.
            val stuff: doNetwork = doNetwork()
            myNet = Thread(stuff)
            myNet!!.start()
        }

        //What is our IP address?
        val wm = getSystemService(WIFI_SERVICE) as WifiManager
        //noinspection deprecation    wifi can't return a ipv6, which is what the issue is, formater doesn't support ipv6
        @Suppress("DEPRECATION")
        val ip = Formatter.formatIpAddress(wm.connectionInfo.ipAddress)
        binding.logger.append("Server IP address is $ip\n")
    }

    /**
     * simple hepler method to update the logger/output textview.
     */
    fun mkmsg(str: String) {
        //so the client uses a handler, while the server code is going to use runUI method.
        //the method name makes no sense here, but keeping the name, so it matches the tpcclient.
        runOnUiThread { binding.logger.append(str) }
    }

    /**
     * Most of the work is done here, so it does not lock the UI thread.  Calls mkmsg, which run on the main thread to do updates.
     *
     * protocol for the example is the server reads in a line and then sends a line.
     * then it closes the connection.
     */
    private inner class doNetwork : Runnable {
        override fun run() {
            val p = binding.ETport.text.toString().toInt()
            mkmsg(" Port is $p\n")
            try {
                mkmsg("Waiting on Connecting...\n")
                Log.v(TAG, "S: Connecting...")
                val serverSocket = ServerSocket(p)

                //socket created, now wait for a coonection via accept.
                val client = serverSocket.accept()
                Log.v(TAG, "S: Receiving...")
                try {
                    //setup send/receive streams.
                    val netIn =
                        BufferedReader(InputStreamReader(client.getInputStream()))  //in is a reserved word in kotlin.
                    val netOut = PrintWriter(
                        BufferedWriter(OutputStreamWriter(client.getOutputStream())),
                        true
                    )

                    //receive the message first.
                    mkmsg("Attempting to receive a message ...\n")
                    val str = netIn.readLine()
                    mkmsg("received a message:\n$str\n")

                    //now send a message.
                    val message = "Hello from server android emulator"
                    mkmsg("Attempting to send message ...\n")
                    netOut.println(message)
                    mkmsg("Message sent...\n")

                    //now close down the send/receive streams.
                    netIn.close()
                    netOut.close()
                } catch (e: Exception) {
                    mkmsg("Error happened sending/receiving\n")
                } finally {
                    mkmsg("We are done, closing connection\n")
                    client.close() //close the client connection
                    serverSocket.close() //finally close down the server side as well.
                }
            } catch (e: Exception) {
                mkmsg("Unable to connect...\n")
                e.printStackTrace()
            }
        }
    }
}